ai-integrations/langchain-parent-document-retrieval.ipynb

```diff
@@ -0,0 +1,188 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "os.environ[\"OPENAI_API_KEY\"] = \"<api-key>\"\n",
+    "ATLAS_CONNECTION_STRING = \"<connection-string>\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
+    "from langchain_community.document_loaders import PyPDFLoader\n",
+    "\n",
+    "# Load the PDF\n",
+    "loader = PyPDFLoader(\"https://investors.mongodb.com/node/12881/pdf\") \n",
+    "data = loader.load()\n",
+    "\n",
+    "# Chunk into parent documents\n",
+    "parent_splitter = RecursiveCharacterTextSplitter(chunk_size=2000, chunk_overlap=20)\n",
+    "docs = parent_splitter.split_documents(data)\n",
+    "\n",
+    "# Print a document\n",
+    "docs[0]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_mongodb.retrievers import MongoDBAtlasParentDocumentRetriever\n",
+    "from langchain_openai import OpenAIEmbeddings\n",
+    "\n",
+    "# Define the embedding model to use\n",
+    "embedding_model = OpenAIEmbeddings(model=\"text-embedding-3-small\")\n",
+    "\n",
+    "# Define the chunking method for the child documents\n",
+    "child_splitter = RecursiveCharacterTextSplitter(chunk_size=200, chunk_overlap=20)\n",
+    "\n",
+    "# Database and collection name\n",
+    "database_name = \"langchain_db\"\n",
+    "collection_name = \"parent_document\"\n",
+    "\n",
+    "# Create the parent document retriever\n",
+    "parent_doc_retriever = MongoDBAtlasParentDocumentRetriever.from_connection_string(\n",
+    "    connection_string = ATLAS_CONNECTION_STRING,\n",
+    "    child_splitter = child_splitter,\n",
+    "    embedding_model = embedding_model,\n",
+    "    database_name = database_name,\n",
+    "    collection_name = collection_name,\n",
+    "    text_key = \"page_content\",\n",
+    "    relevance_score_fn = \"dotProduct\",\n",
+    "    search_kwargs = { \"k\": 10 },\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Ingest the documents into Atlas\n",
+    "parent_doc_retriever.add_documents(docs)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Get the vector store instance from the retriever\n",
+    "vector_store = parent_doc_retriever.vectorstore\n",
+    "\n",
+    "# Use helper method to create the vector search index\n",
+    "vector_store.create_vector_search_index(\n",
+    "   dimensions = 1536\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Run a vector search query\n",
+    "parent_doc_retriever.invoke(\"AI technology\")"
```

Comment:
When I ran this the first time, I didn't get any output. I realized afterward it was because the search index hadn't completed building yet. Since there does not appear to be any easily available method to confirm the search index has built successfully, I wonder if we should add a comment above where we create the index, or here where we're using it to run a query, explaining that you have to wait until the index has completed building.

Another option might be to just add a "sleep" for 60 seconds after creating the index to ensure someone doesn't just execute the entire notebook and then get annoyed when it doesn't work.

> import time
>
> time.sleep(60)